server/session/cookie_splitter.ts

```diff
@@ -0,0 +1,137 @@
+/*
+ *   Copyright OpenSearch Contributors
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License").
+ *   You may not use this file except in compliance with the License.
+ *   A copy of the License is located at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   or in the "license" file accompanying this file. This file is distributed
+ *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *   express or implied. See the License for the specific language governing
+ *   permissions and limitations under the License.
+ */
+import { ensureRawRequest } from '../../../../src/core/server/http/router';
+import { deflateValue, inflateValue } from '../utils/compression';
+
+export interface ExtraAuthStorageOptions {
+  cookiePrefix: string;
+  additionalCookies: number;
+}
+
+export function getExtraAuthStorageValue(request: any, options: ExtraAuthStorageOptions): string {
+  let compressedContent = '';
+  let content = '';
+
+  if (options.additionalCookies > 0) {
+    compressedContent = unsplitCookiesIntoValue(
+      request,
+      options.cookiePrefix,
+      options.additionalCookies
+    );
+  }
+
+  try {
+    content = inflateValue(Buffer.from(compressedContent, 'base64')).toString();
+  } catch (error) {
+    throw error;
+  }
+
+  return content;
+}
+
+/**
+ * Compress and split up the given value into multiple cookies
+ * @param request
+ * @param cookie
+ * @param options
+ */
+export function setExtraAuthStorage(
+  request: any,
+  content: string,
+  options: ExtraAuthStorageOptions
+): void {
+  if (!options.additionalCookies || options.additionalCookies === 0) {
+    return;
+  }
+
+  const compressedAuthorizationHeaderValue: Buffer = deflateValue(content);
+  const compressedContent = compressedAuthorizationHeaderValue.toString('base64');
+
+  if (options.additionalCookies > 0) {
+    splitValueIntoCookies(
+      request,
+      options.cookiePrefix,
+      compressedContent,
+      options.additionalCookies
+    );
+  }
+}
+
+export function splitValueIntoCookies(
+  request: any, // @todo Should be OpenSearchDashboardsRequest, I believe?
+  cookiePrefix: string,
+  value: string,
+  additionalCookies: number
+): void {
+  /**
+   * Assume that Iron adds around 50%.
+   * Remember that an empty cookie is around 30 bytes
+   */
+  const maxLengthPerCookie = Math.floor(4000 / 1.5);
+  const cookiesNeeded = value.length / maxLengthPerCookie; // Assume 1 bit per character since this value is encoded
+
+  // If the amount of additional cookies aren't enough for our logic, we try to write the value anyway
+  // TODO We could also consider throwing an error, since a failed cookie leads to weird redirects.
```

Comment:
It is safe to return `500` or an equivalent response in such cases and fail the ACS request sent by IDP to OSD. 
Not sure about OpenID behaviour in this case.

Reply:
> 3. The auth code redirects back to the IdP

This won't happen it will fail with 400 Bad Request in case of SP Initiated login. 
In case of IdP initiated login we can return 500 and stop the request flow